### PR TITLE
Fix batch_first paths in DPMultiheadAttention

### LINKED ISSUE #774: DPMultiheadAttention Doesn't check attention mask correctly when batch_first is set to True.
## 🐛 Bug

This bug happens when using `DPMultiheadAttention` with the `batch_first` parameter set to True. While going thruough the shape checks, the forward method of `DPMultiheadAttention` check whether the provided mask two last dimensions are `(query.shape(0), key.shape(0))`. This is only correct when batch_first=False.

## To Reproduce
[COLLAB LINK](https://colab.research.google.com/drive/19HriI3H1_k4P8JrcYax2L3Z-ckWp7uI-?usp=sharing)

1. Build a model with a Transformer Encoder that uses `DPMultiheadAttention` as a replacement for vanilla multi-head attention and the batch_first parameter set to True.
2. Make a forward pass with an attention mask.
3. Watch the forward pass fail on the line where it checks the coherence between attention mask shape and query/key first dimension size.

Check the collab example above where I reproduce the error, and show that a patch solves it.

## Expected behavior
I expect the mask size to be matched to the dimension 0 of query/key when batch_first=False and to dimension1  of query/key when batch_first=True.

## Environment

PyTorch version: 2.7.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.11.11 (main, Dec  4 2024, 08:55:07) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.6.87.2-microsoft-standard-WSL2-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               16
On-line CPU(s) list:                  0-15
Vendor ID:                            GenuineIntel
Model name:                           12th Gen Intel(R) Core(TM) i7-12650H
CPU family:                           6
Model:                                154
Thread(s) per core:                   2
Core(s) per socket:                   8
Socket(s):                            1
Stepping:                             3
BogoMIPS:                             5376.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves avx_vnni vnmi umip waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize flush_l1d arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            384 KiB (8 instances)
L1i cache:                            256 KiB (8 instances)
L2 cache:                             10 MiB (8 instances)
L3 cache:                             24 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Vulnerable: No microcode
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.7.0
[pip3] triton==3.3.0
[conda] Could not collect

### PR DESCRIPTION
**Pull request by immu4989**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue

Fixes #774 (plus a closely-related `batch_first=True` regression that surfaced while adding the requested test coverage).

`DPMultiheadAttention.forward()` had two `batch_first`-related issues:

**1. `attn_mask` shape validation ignored `batch_first` (#774).** The check used `query.size(0)` / `key.size(0)` regardless of `batch_first`. When `batch_first=True` those return the batch size, so any valid 2D or 3D mask was incorrectly rejected. Fix: compute `tgt_len` and a parallel `key_seq_len` in a `batch_first`-aware way and reference them in the mask validation.

**2. Output reshape under `batch_first=True` permuted nothing.** The reshape called `.view(bsz, tgt_len, embed_dim)` directly on the `(bsz * num_heads, tgt_len, head_dim)` tensor without first transposing the head and sequence dims, so head positions got spliced into time positions — empirically `max|diff|` against `nn.MultiheadAttention` ≈ 0.96. Fix: explicit `(B*H, L, D) → (B, H, L, D) → (B, L, H, D) → (B, L, E)` via `view → transpose(1, 2) → contiguous → view`. The `batch_first=False` branch is unchanged.

Neither bug surfaced previously because the existing `test_attn` / `test_dp_attn` only exercise `batch_first=False` and never pass `attn_mask`. See #774 for the diagnostic discussion — @iden-kalemaj greenlit bundling both fixes here.

## How Has This Been Tested

Added `test_attn_mask_batch_first` (Hypothesis-based, matching the existing `test_attn` / `test_dp_attn` style) covering both `batch_first` modes and both 2D and 3D mask shapes, asserting output and gradient parity with `nn.MultiheadAttention`.

Verified locally on Python 3.12 / pytest 9.0.3:

- `opacus/tests/dp_layers/dp_multihead_attention_test.py` — 4 tests pass (3 existing + new `test_attn_mask_batch_first`)
- `opacus/tests/grad_samples/dp_multihead_attention_test.py` — passes
- `opacus/tests/validators/multihead_attention_test.py` — both validator tests pass

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Review by iden-kalemaj — APPROVED** [review]
Review automatically exported from Phabricator review in Meta.

**Comment by iden-kalemaj** [comment]
thank you for the fix @immu4989, I have accepted and landing this.
